Document Volume and Claim Pre-binding feature (#22946)
kubernetes/website: content/en/docs/concepts/storage/persistent-volumes.md

However, the particular path specified in the custom recycler Pod template in the `volumes` part is replaced with the particular path of the volume that is being recycled.

### Pre-binding

> The control plane can [bind Persistent Volume Claims to matching Persistent Volumes](#binding) in the

Reviewer: @kbhawkey  we use PVC + PV throughout this doc. Do you think it's still better to spell out the abbreviation when we already introduced it earlier?

Reviewer 2: Hi. Sorry for the confusion. I reread the paragraphs. When the terms are spelled out, it helps me understand the
concepts, especially when there is a number of cross referencing of fields in the configuration files. I am not against abbreviating the terms. The abbreviations are defined at the top of the document
and used throughout the page.

Is "Pre-binding" the correct heading or would "Preferred binding" or "Selective binding" make more sense?

Reviewer: How about something like "Reserve a PersistentVolume for a PersistentVolumeClaim"?

In terms of abbreviation vs not, I'm fine either way if you think it's easier to understand. What do you think of keeping PersistentVolume and PersistentVolumeClaim as one word to match the object name?

Reviewer 2: OK. The new section is found under the parent section, "Lifecycle of a volume and claim".
The other sub-section headings are one word.
One idea: Reserving a PersistentVolume?

Reviewer: I like "Reserving a PersistentVolume"

Author: Updated.

> The control plane can [bind Persistent Volume Claims to matching Persistent Volumes](#binding) in the

Reviewer 3: ```suggestion
The control plane can [bind PersistentVolumeClaims to matching PersistentVolumes](#binding) in the
```

> cluster. However, if you want a Persistent Volume Claim to bind to a specific Persistent Volume, you need to pre-bind them.

Reviewer 3: ```suggestion
cluster. However, if you want a PVC to bind to a specific PV, you need to pre-bind them.
```

> You can specify the Persistent Volume in your Persistent Volume Claim, then the Persistent Volume Claim will be able to bind to this specific Persistent Volume. If the Persistent Volume exists and is not reserved for other Persistent Volume Claims (has Persistent Volume Claim specified in `claimRef` field), the Persistent Volume and Persistent Volume Claim will be bound regardless of whether the Persistent Volume satisfies some matching criteria, e.g. node affinity. Storage class, access modes, and requested storage size are still checked.

Reviewer: The last part of the sentence is a little confusing. It says it will be bound regardless of matching criteria, but then we have a list of criteria that is still checked.

Reviewer 2: Does this help:

By setting a PV in a Persistent Volume Claim, you declare a binding between a specific Persistent Volume and a PVC.
If the Persistent Volume exists and has not reserved PersistentVolumeClaims through its `claimRef` field, then the Persistent Volume and Persistent Volume Claim will be bound regardless of whether the Persistent Volume satisfies some matching criteria such as node affinity. Storage class, access modes, and requested storage size are still checked.

Can the binding fail depending on the outcome of the items in the last sentence?

Reviewer: Ah sorry. I missed the period after affinity. This makes a lot more sense now.

Reviewer 3: ```suggestion
By specifying a PersistentVolume in a PersistentVolumeClaim, you declare a binding between that specific PV and PVC.
If the PersistentVolume exists and has not reserved PersistentVolumeClaims through its `claimRef` field, then the PersistentVolume and PersistentVolumeClaim will be bound.

The binding happens regardless of some volume matching criteria, including node affinity.
The control plane still checks that [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/), access modes, and requested storage size are valid.
```
?

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: foo-pvc
  namespace: foo
spec:
  volumeName: foo-pv
  ...
```

> This method does not guarantee any binding privileges to the Persistent Volume. If your specified Persistent Volume can be used by other Persistent Volume Claims, you need to reserve it first by specifying the Persistent Volume Claim in the Persistent Volume using the `claimRef` field so that other Persistent Volume Claims can not bind to it.

Reviewer 3: ```suggestion
This method does not guarantee any binding privileges to the PersistentVolume. If other PersistentVolumeClaims could use the PV that you specify, you first need to reserve that storage volume. List the relevant PersistentVolumeClaim in the `claimRef` field of the PV so that other PVCs can not bind to it.
```

```
apiVersion: v1

> kind: PersistentVolum

Reviewer 2: nit: kind: PersistentVolume

> kind: PersistentVolum

Reviewer 3: ```suggestion
kind: PersistentVolume
```

However, the particular path specified in the custom recycler Pod template in the `volumes` part is replaced with the particular path of the volume that is being recycled.

### Reserving a PersistentVolume

The control plane can [bind PersistentVolumeClaims to matching PersistentVolumes](#binding) in the
cluster. However, if you want a PVC to bind to a specific PV, you need to pre-bind them.

By specifying a PersistentVolume in a PersistentVolumeClaim, you declare a binding between that specific PV and PVC.
If the PersistentVolume exists and has not reserved PersistentVolumeClaims through its `claimRef` field, then the PersistentVolume and PersistentVolumeClaim will be bound.

The binding happens regardless of some volume matching criteria, including node affinity.
The control plane still checks that [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/), access modes, and requested storage size are valid.

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: foo-pvc
  namespace: foo
spec:
  volumeName: foo-pv
  ...
```

> This method does not guarantee any binding privileges to the PersistentVolume. If other PersistentVolumeClaims could use the PV that you specify, you first need to reserve that storage volume. List the relevant PersistentVolumeClaim in the `claimRef` field of the PV so that other PVCs can not bind to it.

Reviewer: Nit: List => Specify.

However, the particular path specified in the custom recycler Pod template in the `volumes` part is replaced with the particular path of the volume that is being recycled.

### Pre-binding

> The control plane can [bind PVCs to matching PVs](#binding) in the

Reviewer 4: ```suggestion
The control plane can [bind persistent volume claims to matching persistent volumes](#binding) in the
```

> cluster. However, if you want a PVC to bind to a specific PV, you need to

Reviewer 4: ```suggestion
cluster. However, if you want a persistent volume claim to bind to a specific persistent volume, you need to
```

pre-bind them.

> You can specify the PV in your PVC, then the PVC will be able to bind to this
> specific PV. If the PV exists and is not reserved for other PVCs (has PVC
> specified in `claimRef` field), the PV and PVC will be bound regardless of
> whether the PV satisfies some matching criteria, e.g. node affinity. Storage
> class, access modes, and requested storage size are still checked.

Reviewer 4: ```suggestion
You can specify the persistent volume in your persistent volume claim, then the persistent volume claim will be able to bind to this
specific persistent volume. If the persistent volume exists and is not reserved for other persistent volume claims (has persistent volume claim
specified in `claimRef` field), the persistent volume and persistent volume claim will be bound regardless of
whether the persistent volume satisfies some matching criteria, e.g. node affinity. Storage
class, access modes, and requested storage size are still checked.
```

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: foo-pvc
  namespace: foo
spec:
  volumeName: foo-pv
  ...
```

> Note that this method does not guarantee the PV right. If the PV you specified

Reviewer 2: Hi @cofyc . Several doc-related questions and suggestions.
nit: You could spell out the terms instead of only using the abbreviations. I find it hard to read about `PV` and `PVC`.
Question: What method? A method described in the previous paragraph? The `pre-binding` method?
nit: Wording, the phrase, "This method does not guarantee any [what?] privileges to the Persistent volume? 
nit: What do you think about adding this information using a `note` shortcode?

nit: If your specified Persistent volume can be used by other Persistent Volume Claims, you need to reserve ...

Reviewer 4: ```suggestion
Note that this method does not guarantee the persistent volume right. If the persistent volume you specified
```

Author: > nit: You could spell out the terms instead of only using the abbreviations. I find it hard to read about PV and PVC.

done, thanks @celestehorgan ! but the abbreviations PV & PVC are widely used in this doc (https://kubernetes.io/docs/concepts/storage/persistent-volumes/). Is it necessary?

> Question: What method? A method described in the previous paragraph? The pre-binding method?

the method described in the previous paragraph

> nit: Wording, the phrase, "This method does not guarantee any binding privileges to the Persistent volume?
> nit: What do you think about adding this information using a note shortcode?
> nit: If your specified Persistent volume can be used by other Persistent Volume Claims, you need to reserve ...

this introduces the reserve approach, I think it's better not to add this information using a note shortcode.

I've updated the PR and fixed the wording too. Thanks! PTAL.

> may be used by other PVCs, you need to reserve it first by specifying the PVC
> in the PV using the `claimRef` field so that other PVCs can not bind to it.

Reviewer 4: ```suggestion
may be used by other persistent volume claims, you need to reserve it first by specifying the persistent volume claim
in the persistent vlume using the `claimRef` field so that other persistent volume claims can not bind to it.
```

However, the particular path specified in the custom recycler Pod template in the `volumes` part is replaced with the particular path of the volume that is being recycled.

### Pre-binding

> The controller in the master can [bind PVCs to matching PVs](#binding) in the
> cluster. However, if you want a PVC to bind to a specific PV, you need to

Reviewer 3: ```suggestion
The control plane can [bind PVCs to matching PVs](#binding) in the
cluster. However, if you want a PVC to bind to a specific PV, you need to
```

Kubernetes itself does not have master nodes (some tools, like kubeadm, did have this concept but are moving to a new name).

Author: Thanks!

pre-bind them. As it's a bi-directional binding between the PVC and the PV,
there are two ways:

> - Specify the PV in your PVC using the `volumeName` field

Reviewer: I believe this first method doesn't guarantee the PV right? You could have:
* 2 pvcs pointing to the same pv
* a second pvc that doesn't specify any pv, but matches it.

Should we just recommend the 2nd way of setting claimRef?

Author: yes, the first method doesn't guarantee.

How about documenting it but we emphasize that it doesn't guarantee the PV right?

Sometimes it's useful as it is simpler and does not require `UPDATE` permission on cluster resource `PV`.

Author: maybe it's better not to say there are two ways.
I made an update. https://github.com/kubernetes/website/pull/22946/commits/839184d4098a648ac5ff14c3b13a8ebc6bc351ea
what do you think about it?

- Specify the PVC in your PV using the `claimRef` field

Normally, you can only specify the PV in your PVC, then the PVC will only be
able to bind to this specific PV. If the PV exists and is `Available`, the PV
and PVC will be bound regardless of whether the PV satisfies the matching
criteria, e.g. node selector, storage request.

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: foo-pvc
  namespace: foo
spec:
  volumeName: foo-pv
  ...
```

If the PV may be used by other PVCs, you need to reserve it by specifying the
PVC in the PV using the `claimRef` field so that other PVCs can not bind to it.

```
apiVersion: v1
kind: PersistentVolum
metadata:
  name: foo-pv
spec:
  claimRef:
    name: foo-pvc
    namespace: foo
  ...
```

> This is useful if you want to consume PVs of which the data is retained before.

Reviewer 3: ```suggestion
This is useful if you want to consume PVs that already exist.
```

Author: there are two scenarios:

1) PVs are statically-provisioned and are empty
2) PVs are statically-provisioned or dynamically-provisioned but are retained (`claimPolicy` is `Retain`), so that they can be reused by other PVCs

I want to emphasize the latter. As for the former, users don't need to pre-bind the PVC and the PV normally.

Reviewer 3: How about:
```suggestion
This is useful if you want to consume PersistentVolumes that have their `claimPolicy` set
to `Retain`, including cases where you are reusing an existing PV.
```
?
